Add syntax highlighting for Mojo

mahmoudajawad opened the pull request:
Mojo is a new language, a superset of Python. Vim currently doesn't support it. This PR adds necessary basic support.

Closes #13062

mahmoudajawad committed:
Add syntax highlighting for Mojo

dkearns commented on runtime/syntax/mojo.vim:
```diff
@@ -0,0 +1,313 @@
+" https://docs.modular.com/mojo/lib.html
+"
+if !exists("mojo_no_builtin_highlight")
+  syn keyword mojoBuiltin	Byte ListLiteral slice constrained CoroutineContext
```
The types here might be better highlighted using the `Type` highlight group.  See `:help group-name`

mahmoudajawad replied:
If `Type` refers to buil-in type then this is definitely the case. I'm updating this.

mahmoudajawad replied:
More on this.
Some of what I defined here are built-in functions such as `debug_assert`. Should I keep these in a `Builtin` group, while everything else that represents a type is changed to `Type`?

dkearns replied:
Yes, I'd create a mojoType syntax group and leave the rest of the builtins in the existing group.

dkearns commented on runtime/syntax/mojo.vim:
```diff
@@ -0,0 +1,313 @@
+syn keyword mojoStatement	lambda nonlocal pass return with yield
+syn keyword mojoStatement	class def nextgroup=mojoFunction skipwhite
+syn keyword mojoStatement	struct fn nextgroup=mojoFunction skipwhite
+syn keyword mojoIdentifier	alias var let
```
This is linked to the Identifier highlight group but that is usually used for the variable name, if at all.  Using Statement for these would be the standard approach.

mahmoudajawad replied:
Are you referring to `alias, var, let`? I found that `typescript` is using `Identifier` and followed that. I got confused myself on this.

dkearns replied:
Yes.  Some of the syntax definitions try and classify the keywords in this manner but it's definitely not the norm.

mahmoudajawad committed:
Add Mojo filetype definition

mahmoudajawad committed:
Add test for Mojo filetype definition

dkearns commented on runtime/filetype.vim:
```diff
@@ -1368,6 +1368,10 @@ au BufNewFile,BufRead *.rc,*.rch
 
+" Mojo
+" Mojo files use either .mojo or .🔥 as extension
+au BufNewFile,BufRead *.mojo,*.🔥		setf python
```
This should be `setf mojo`.

mahmoudajawad committed:
Fix syntax difintion of Mojo files

mahmoudajawad committed:
Update Mojo syntax highlight

Drop group Identifier for alias, var, let

mahmoudajawad committed:
Update Mojo syntax highlight

Re-organize built-ins into 'Builtin' for built-in functions, and 'Type' for built-in types.
